src/app/clusters/level-control/level-control.cpp

```diff
@@ -139,6 +146,34 @@ static void timerCallback(System::Layer *, void * callbackContext)
     emberAfLevelControlClusterServerTickCallback(static_cast<EndpointId>(reinterpret_cast<uintptr_t>(callbackContext)));
 }
 
+static uint32_t computeCallbackWaitTimeMs(CallbackScheduleState & callbackSchedule, uint32_t delayMs)
+{
+    auto delay             = System::Clock::Milliseconds32(delayMs);
+    auto waitTime          = delay;
+    const auto currentTime = System::SystemClock().GetMonotonicTimestamp();
+
+    // Compute the wait time until the next callback based on the previous schedule latency
+    // and on the previous callback duration.
+    if ((currentTime > callbackSchedule.nextIdealTimestamp) && (callbackSchedule.prevDuration < delay))
```

Comment:
My apologies for the lag here; I was out sick, unfortunately.

And I will be out all of next week, so this might need someone else to review.

But I still don't understand the logic here.  This is doing adjustment at the point when currentTime is already past the `nextIdealTimestamp`.  Is that the `nextIdealTimestamp` for the _next_ callback?  Or the _current_ callback?  Why isn't the relevant comparison not how currentTime + waitTime compares to the expected callback timestamp?

And if we _are_ overdue already, why does it matter how `prevDuration` compares to `delay`?

This all needs either a _lot_ more documentation or clearer naming or both.